Reduce number of queries during GraphQL execution

lutter opened the pull request:
For a query like `parents { id children { id } }` we used to run `1 + number(parents)` queries, one to get the parents, and then one query to get the children of each parent. The more deeply a query was nested, the more this effect compounded.

With this PR, we will now run 2 queries: one to get the parents, and one to get all children for the parents. Details of how this is done can be found in [this document](https://github.com/graphprotocol/graph-node/blob/dfefdf775b68db202e8fdf2672a9e909f0b21e59/docs/implementation/query-prefetching.md)

It is possible to have `graph-node` compare the results of execution with and without prefetching by adding a `@verify` directive to the query, as in `query stuff @verify { .. }` (this must use the form with an explicit `query` keyword) In this mode, the query will be executed in the old and the new way; if the two results differ, the query returns an error that contains both result for manual comparison.

It is also possible to completely turn off prefetching by setting the environment variable `GRAPH_GRAPHQL_NO_PREFETCH`; if that variable is set to anything at all, query execution behaves like it did before this PR. (see #1340)

Supersedes PR #1341

lutter commented:
Rebased to latest master

Jannis commented on graphql/src/store/prefetch.rs:
```diff
@@ -0,0 +1,870 @@
+///
+/// That distinguishes it from both the result of a query that matches
+/// nothing (an empty `Vec`), and a result that finds just one entity
+/// (the entity is not completely empty)
```
A cleaner way would be for `Node` to be an enum and to have a `Node::Root` variant for the root.

lutter replied:
I thought about doing that, but it would complicate the code quite a bit, since the root node and interior nodes are treated exactly the same, except for one place where we decide whether we need to window the query or not (in `fetch`) And since the data structure is completely internal to `prefetch`, I don't think it will lead to a lot of confusion.

Jannis replied:
Yeah, I thought so. Just wanted to bring it up.

Jannis commented on store/postgres/tests/store.rs:
```diff
@@ -2289,3 +1991,185 @@ fn handle_large_string_with_index() {
+            .into_iter()
+            .map(|child_type| {
+                let attribute = WindowAttribute::Scalar("favorite_color".to_owned());
+                let link = EntityLink::Direct(attribute);
```
Are we indirectly testing the other `EntityLink` variant or should we add tests for this here as well?

lutter replied:
Yes, they are covered by the tests in `core/tests/interfaces.rs` and `graphql/tests/query.rs`

Jannis replied:
Ah, ok, good.

lutter commented on docs/implementation/query-prefetching.md:
```diff
@@ -0,0 +1,505 @@
+That leaves us with the following combinations of whether the parent and
+child store a list or a scalar value, and whether the parent is derived:
+
+<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">
```
Neat! I didn't know that is possible. Since we now have an RFC repo, I wonder if that doc should really go there - otherwise, it's going to be the only document about implementation in `graph-node`.

Jannis replied:
It's not quite an engineering plan but I agree, it's a good place to put it. Could you create an engineering plan PR with this document as `PLAN-0001: SQL query combination`?

lutter commented on docs/implementation/query-prefetching.md:
```diff
@@ -0,0 +1,505 @@
+
+### Type A
+
+Use when parent is derived and child is a list (_tt)
```
It shows which part of the table above each case covers, i.e. `_tt` is for "doesn't matter if parent is a list; parent is derived; child is a list" I mostly put it there to make sure I covered all possibilities.

Jannis replied:
I think I still don't know what they stand for? `_` is don't care, `t` and `f` are... what? 😁

Perhaps we can put something human-readable in the parentheses there?

lutter replied:
`t` is true and `f` is false, but I'll just take it out, since it just repeats what the sentence says in more concise form.

lutter commented on store/postgres/src/relational_queries.rs:
```diff
@@ -140,7 +142,10 @@ impl EntityData {
                     // column; those will be things like the block_range that
                     // is used internally for versioning
-                    if let Some(column) = table.column(&SqlName::from_snake_case(key)).ok() {
+                    if key == "parent_id" {
```
Postgres actually allows `$` in identifiers, which is illegal in GraphQL names; I'll just change it to `g$parent_id` (`g` for 'graph) That way, it's impossible to collide with a GraphQL name.

lutter replied:
I also renamed `pos` in the queries to `g$pos` to avoid possible collisions.

lutter commented on store/postgres/src/relational_queries.rs:
```diff
@@ -1027,150 +1284,367 @@ impl<'a> FilterQuery<'a> {
+
+    // Generate (taking optionality of either clause into account)
+    //    where {pos} > {order.skip} and {pos} <= {order.first + order.skip}
+    // `pos` must be the name of the column we order by and 'conj' must be
```
This is explaining what the `pos` parameter of `limit_per_window` is for; but I just noticed that that is always `c.pos`, so we can get rid of that parameter.

lutter commented on graphql/src/store/resolver.rs:
```diff
@@ -154,22 +156,110 @@ where
+    ) -> Option<&'a q::Value> {
+        match parent {
+            Some(q::Value::Object(map)) => {
+                let key = format!("r:{}", qast::get_response_key(field));
```
No specific reason; I just changed it to `prefetch:`

lutter commented on graph/src/data/query/error.rs:
```diff
@@ -54,6 +55,11 @@ pub enum QueryExecutionError {
+    IncorrectPrefetchResult {
+        single: q::Value,
+        prefetch: q::Value,
+    },
```
Renamed it.

lutter commented:
Addressed all review comments, except for where to put the implementation notes for all this. I think they really belong in the RFC repo now as that will contain similar material in the future.

Jannis commented on docs/implementation/query-prefetching.md:
```diff
@@ -0,0 +1,505 @@
+     order by parent_id, pos
+
+When there is only one window, we can simplify the above query. The
+simplifcation basically inlines the `matches` CTE. That is important as
```
This typo hasn't been fixed yet.

lutter replied:
I didn't push updates to the doc since I was waiting whether you agree to put it in the RFC repo

lutter commented:
Removed the implementation details doc and opened https://github.com/graphprotocol/rfcs/pull/2 for it. That should address all the comments in the review.

lutter commented:
Rebased to latest master
